File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/RequireDocContentRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.stream.Stream;

/**
 * Rule that requires specific content in documentation files.
 */
public enum RequireDocContentRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireDocContent";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final JsonNode checksNode = catNode.get("parameters").get("checks");
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return Stream.of(checksNode.spliterator(), false)
                .flatMap(check -> validateCheck(root, check, severity).stream())
                .toList();
    }

    private List<Finding> validateCheck(Path root, JsonNode check, String severity) throws MojoExecutionException {
        final List<String> containsAll = HarnessCheckHelper.extractPaths(check.get("containsAll"));
        final String failureMessage = check.get("failureMessage").asText();
        final String combined = HarnessCheckHelper.extractPaths(check.get("files")).stream()
                .map(f -> {
                    try {
                        return HarnessCheckHelper.readFile(root, root.resolve(f));
                    } catch (MojoExecutionException e) {
                        return "";
                    }
                })
                .reduce((a, b) -> a + "\n" + b)
                .orElse("");
        return containsAll.stream().allMatch(combined::contains)
                ? List.of()
                : List.of(new Finding(severity, CATEGORY, failureMessage));
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireDocContentRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessPsiResults
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path

/**
 * Rule that requires documentation files to contain specified content.
 */
object RequireDocContentRule : HarnessCheckRule {
    override fun applies(manifest: JsonObject): Boolean {
        val category = "requireDocContent"
        val catObj = manifest[category]?.jsonObject ?: return false
        val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
        return enabled
    }

    override fun validate(
        manifest: JsonObject,
        root: Path,
        psiResults: HarnessPsiResults?,
    ): Collection<Finding> {
        val category = "requireDocContent"
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val checks = parametersObj?.get("checks")?.jsonArray
        return if (catObj == null || parametersObj == null || checks == null) {
            emptyList()
        } else {
            checks
                .filter { checkElem ->
                    val checkObj = checkElem.jsonObject
                    val files = HarnessCheck.stringArrayFrom(checkObj, "files")
                    val containsAll = HarnessCheck.stringArrayFrom(checkObj, "containsAll")
                    val content = files.map { filePath -> HarnessCheck.readSafe(root, filePath) }.joinToString("\n")
                    !containsAll.all { contentStr -> content.contains(contentStr) }
                }.map { checkElem ->
                    val checkObj = checkElem.jsonObject
                    Finding(
                        HarnessCheck.severityOf(manifest, category),
                        category,
                        HarnessCheck.stringFrom(checkObj, "failureMessage"),
                    )
                }
        }
    }
}
